apm-agent-core/src/main/java/co/elastic/apm/agent/context/LifecycleListener.java

```diff
@@ -32,14 +32,19 @@
  * Implement this interface and register it as a {@linkplain java.util.ServiceLoader service} under
  * {@code src/main/resources/META-INF/services/co.elastic.apm.agent.context.LifecycleListener}.
  * </p>
+ * <p>
+ * Implementations may have a constructor with an {@link ElasticApmTracer} argument
+ * </p>
  */
 public interface LifecycleListener {
 
     /**
      * Callback for when the {@link ElasticApmTracer} starts.
      *
      * @param tracer The tracer.
+     * @deprecated in favor of a constructor with an {@link ElasticApmTracer} argument
      */
+    @Deprecated
```

Comment:
After thinking a bit more about this, I'm not sure that deprecating the `start` method and replace with constructor arguments is the perfect strategy here. As I might be close to bike-shedding I'll elaborate a bit:
- on the positive side, it allows to keep a final reference to `ElasticApmTracer` instance instead of having a mutable field
- on the other side, there is no enforcement at compile time for implementing any constructor contract, thus it's likely to produce errors at runtime when creating instances of such classes.
- if listener needs to start a thread (or do any kind of expensive operation) when starting, doing it in the constructor is generally not the recommended option, and keeping a `start` would probably provide better semantics in this case.

Reply:
I agree. Let's not deprecate it but still make it possible to inject the tracer into the constructor so implementations can choose which is more appropriate for them. For the `JmxMetricTracker` not only the tracer instance variable can't be `final` but because the `ChangeListener` is invoked from another thread, it would also have to be `volatile`. Thinking about it, we must not register the listener within the constructor as there's a chance that the listener gets invoked before the constructor finishes. Another reason to keep both the constructor tracer argument and the `start` method.